### PR TITLE
Add CodeMirror to the HTML block

### PR DESCRIPTION
**Pull request by noisysocks**
### 👀 What this is

Adds CodeMirror to the HTML block. This means that a user can type their custom HTML using a fully featured text editor with linting and syntax highlighting:

codemirror

### 🤔 How this works

[CodeMirror was added to core in WordPress 4.9](https://make.wordpress.org/core/2017/10/22/code-editing-improvements-in-wordpress-4-9/) which means we can enhance a `<textarea>` simply by calling `wp.codeEditor.initialize()`.

To make CodeMirror easier to use within React, I introduced a `<CodeEditor>` component which acts as a drop-in replacement for a `<textarea>`.

I've also incorporated some of @westonruter's work in https://github.com/WordPress/gutenberg/pull/1625 by using `SandBox` within our HTML block. This makes previewing a HTML block more robust. Specifically, you can now insert HTML that blocks rendering, e.g.:

```html
<script>document.write( 'Hey!' );</script>
```

To avoiding adding a whopping 1.9 MB to our assets, I've implemented a very simple mechanism to lazily load the assets that are normally queued by calling `wp_enqueue_code_editor()`. 

### ✅ How to test

1. Create a HTML block
2. Preview it
3. Save it, view the page
4. Check that you can preview HTML that blocks rendering, e.g. `<script>document.write( 'Hey!' );</script>`
4. Log in as an author and test that you get lint warnings when using e.g. the `<script>` tag

### CODE REVIEW

**Comment by gziolo** [comment]
Cross-posting related issue: #2768. I didn't take any measurements on the size of CodeMirror, but I would expect it isn't a small one :) I think it's fine to include it temporarily on the initial editor load. However we should at least double check what impact it has on the size of JS and CSS loaded, and mention it in the #2768 so we know if we have to act before it gets merged into the core.

**Comment by noisysocks** [comment]
Here are some quick measurements from running `npm run build` and checking DevTools:

| Branch | All | JS | CSS |
| --- | --- | --- | --- |
| `master` | 3.4 MB | 2.9 MB | 261 KB |
| This PR | 5.3 MB | 4.8 MB | 277 KB |

Yikes! A 1.9 MB (55%) increase! 😱 I'll definitely look into dynamically loading the assets.

**Comment by noisysocks** [comment]
cc'ing some people who might want to track this work:

- @mtias — suggested it
- @jasmussen — mocked it up in #1386 
- @westonruter — worked on adding CodeMirror to WP Core
- @dmsnell — worked on adding a CodeMirror block to Gutenberg in #1625

**Comment by gziolo** [comment]
> A 1.9 MB (55%) increase! 😱 

@noisysocks thanks for checking 🙇 

Another thing we can add as a follow-up PR is HTML linting which is built-in in the core. By the way, the code for linting is lazy-loaded when editing plugin files. Which is a very good sign since we discovered that CodeMirror adds 1.9 MB to the uncompressed build size.

File: components/code-editor/test/index.js
Comment (gziolo) [inline components/code-editor/test/index.js]: We could use snapshot testing here, too. See example: https://github.com/WordPress/gutenberg/blob/300ca0a8acc73ff4013c0495a8c636527fcf588f/components/notice/test/index.js#L15 and docs: https://github.com/WordPress/gutenberg/blob/master/docs/testing-overview.md#snapshot-testing.

**Comment by mtias** [comment]
> A 1.9 MB (55%) increase! 😱

Ouch.

File: components/code-editor/test/index.js
Comment (noisysocks) [reply components/code-editor/test/index.js]: Good idea. I've changed this to a snapshot test in 17dbeb231a901007b08e5fd0dd75840bdbbc1489.

**Comment by noisysocks** [comment]
Any feedback, questions, or concerns with this PR?

I'd like to get the basic functionality merged and then add lazy loading in a seperate PR. I've started this work over in https://github.com/WordPress/gutenberg/pull/4500.

**Comment by jasmussen** [comment]
Really nice, work, fun to see this, it's a very very cool block.

There are a couple of things we should address in this PR. First off, can we tweak the padding a little bit so it looks like this?

screen shot 2018-01-16 at 13 36 08

I think this should do it:

```
.CodeMirror-lines {
padding: 8px 8px 8px 0;
}
```

There's a z-index issue with adjecent block toolbars:

screen shot 2018-01-16 at 13 33 12

One important issue we have to fix is that you need to be able to escape this block using just the keyboard:

- Pressing Escape while you're editing inside the block should deselect the entire block, just like it does in a Paragraph block
- Pressing the ↓ at the end of a codemirror block should move the caret into the next adjecent block, same with ↑ at the beginning of a block
- _Ideally_ you'd also be able to press Enter twice at the end of a codemirror block to create a new Paragraph block below, just like how you can press Enter twice inside a list to exit it

Question for a future PR, doesn't have to be this one — can we customize the colors of the syntax highlighting? If yes, I'd love for us to use more WordPressy colors. Alternatively, I'd love it if we could at least use something like `$blue-medium-200` for the line-highlighter.

**Comment by noisysocks** [comment]
> Question for a future PR, doesn't have to be this one — can we customize the colors of the syntax highlighting?

Yep! It's just a matter of styling the `.CodeMirror` classes 🙂

**Comment by noisysocks** [comment]
Thanks @jasmussen! I've implemented your suggested changes.

> Pressing the ↓ at the end of a codemirror block should move the caret into the next adjecent block, same with ↑ at the beginning of a block

I made it so that pressing UP/DOWN first moves the cursor to the start/end of the line. A subsequent press will then move focus to the previous/next block. It feels really natural in my testing.

> Ideally you'd also be able to press Enter twice at the end of a codemirror block to create a new Paragraph block below, just like how you can press Enter twice inside a list to exit it

I didn't implement this since pressing ENTER twice is quite common while writing code and e.g. adding whitespace between lines.

**Comment by dmsnell** [comment]
sorry I'm coming in somewhat late. what did we decide about async-loading CodeMirror? are we wanting to still do that?

**Comment by westonruter** [comment]
> sorry I'm coming in somewhat late. what did we decide about async-loading CodeMirror? are we wanting to still do that?

Yes, see https://github.com/WordPress/gutenberg/pull/4500

**Comment by noisysocks** [comment]
Any more thoughts on this? 😊

**Comment by dmsnell** [comment]
I'd really like to see us wait until this is loading on-demand before merging it otherwise it's going to instantly bloat the editor in no small way - that is, I'd love to see something like #4500 merge first _or_ have some changes here to lazy-load.

**Comment by noisysocks** [comment]
I'll hold off on hitting the merge button until #4500 is ready. Nonetheless, I'd like to get this PR into an approved and ready-to-go state 🙂

**Comment by jasmussen** [comment]
Visually this is in a good place 👍:

screen shot 2018-01-23 at 09 05 26

I also appreciate that the arrow at the beginning and end of the block enters adjecent blocks or creates a new paragraph so at least you can escape the block. 👍 

However I'm seeing some weird behavior when testing. Not sure if it's intentional, but basic dummy text paragraphs are joined together on reload:

codeblock

**Comment by noisysocks** [comment]
> However I'm seeing some weird behavior when testing. Not sure if it's intentional, but basic dummy text paragraphs are joined together on reload:

Good catch! Looks like this is because Gutenberg pretty prints the post HTML before saving. Since this bug exists on `master` as well, I'll create an issue for it.

**Comment by noisysocks** [comment]
As per the discussion over in https://github.com/WordPress/gutenberg/pull/4500, I've implemented lazy loading in a one-off manner in e51f342. The 1.9 MB of CodeMirror scripts and styles are now not loaded until a HTML block is instantiated.

**Comment by noisysocks** [comment]
Rebased and fixed conflicts. How do we feel about getting this in? cc. @westonruter @mtias

File: components/code-editor/README.md
Comment (westonruter) [inline components/code-editor/README.md]: Attributes here have inconsistent indentation.

File: components/code-editor/README.md
Comment (westonruter) [inline components/code-editor/README.md]: `s/highliting/highlighting/`

File: components/code-editor/editor.js
Comment (westonruter) [inline components/code-editor/editor.js]: I love this.

**Review by westonruter — APPROVED** [review]
A couple minor things I commented on.

Also, there is a bit of an unexpected UX behavior whereby if I shift-arrow to select some text, I get the block toolbar appearing. I understand that normally makes sense because toolbar buttons like bold and italic. But since there are no such buttons in the toolbar for Custom HTML (yet, as I suppose there could be shortcuts to add common tags), it is a little distracting to see the HTML/Preview buttons flash whenever I select some text with the keyboard.

**Comment by westonruter** [comment]
Per my review comment, here's me with cursor in block:

image

And then as soon as I select some text with keyboard:

image

It doesn't feel like the HTML/Preview buttons show appear.

File: components/code-editor/README.md
Comment (noisysocks) [reply components/code-editor/README.md]: Oops. Fixed in 87af208e500b0dbde3bcb68b0b846a1379692267.

File: components/code-editor/README.md
Comment (noisysocks) [reply components/code-editor/README.md]: Oops. Fixed in 87af208e500b0dbde3bcb68b0b846a1379692267.

**Comment by noisysocks** [comment]
Thanks for reviewing this, @westonruter!

> It doesn't feel like the HTML/Preview buttons show appear.

Good catch. I agree that it feels weird. I think to address this we need to split `<Slot name="Block.Toolbar">` into two slots: one for toolbar buttons that affect the block and one for toolbar buttons that affect the selected text. Only the latter should be shown when text is selected.

Since this would require moving a few things around and it's not a major issue, I'll defer this work for later.

**Pull request merged**
